Implement UNION ALL statement

mxm opened the pull request:
This PR introduces support for the UNION ALL statement with the following commits:

1. Changes to the Analyzer
Most notably `UnionSelect` (to encapsulate the union operation) and `OrderedLimitedRelation` (to represent the relations on both sides of the union.

2. Changes to the Planner
Most notably the bucket ids for the `PageBucketReceiver`s are not strictly sequential anymore. Instead, the bucket id itself is prepended with the branch id. This enables to branch off operations which have the same return type (like in UNION) in the execution tree and re-use the `PageDownstreamContext` without having a dedicated `DownstreamExecutionSubContext`.

3. Entry in the SQL features table
Support for only UNION ALL has been added (excluding the DISTINCT variant)
 
4. User documentation including a doc test

seut commented on sql/src/main/java/io/crate/planner/operators/Count.java:
```diff
@@ -89,10 +89,11 @@ public ExecutionPlan build(PlannerContext plannerContext,
             Collections.singletonList(DataTypes.LONG),
             Collections.singletonList(MergeCountProjection.INSTANCE),
-            DistributionInfo.DEFAULT_SAME_NODE,
+            DistributionInfo.DEFAULT_BROADCAST,
```
also here, why needed? afaik the default should stay same node, only if used intermediate it should change.

mxm replied:
Good point. No need to change the default here. Instead, we should adapt the distribution info in case it cannot be optimized same-node.

seut commented on sql/src/main/java/io/crate/planner/operators/Union.java:
```diff
@@ -0,0 +1,218 @@
+            Collections.singletonList(plannerContext.handlerNode()),
+            leftResultDesc.streamOutputs(),
+            Collections.emptyList(),
+            DistributionInfo.DEFAULT_BROADCAST,
```
this phase runs on the handler node, so distribution should be same_node

seut commented on sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java:
```diff
@@ -0,0 +1,289 @@
+import static org.hamcrest.Matchers.arrayContainingInAnyOrder;
+import static org.hamcrest.Matchers.is;
+
+@ESIntegTestCase.ClusterScope(minNumDataNodes = 1, maxNumDataNodes = 5)
```
is restricting the number of nodes here needed?

mxm replied:
Not the upper bound but we also want to ensure execution with one node occasionally.

mxm commented on sql/src/main/java/io/crate/analyze/relations/UnionSelect.java:
```diff
@@ -0,0 +1,117 @@
+        querySpec.outputs(new ArrayList<>(fields.asList()));
+    }
+
+    public boolean isDistinct() {
```
I'd say because we implemented parser support for the union and have this information available anyway. You're right the analyzer doesn't really need to know that yet. I'd be ok with removing it.

matriv replied:
I agree with @seut to remove it since as he pointed out we don't even use it in toString() currently. So when we implement the UNION [DISTINCT] we don't forget this for example.

mxm commented:
Thanks for the review @seut. I've reverted back the changes where I changed `DistributionInfo` from SAME_NODE to BROADCAST. That was not correct. In fact it turns out the same node optimization is currently not possible with the union implementation. I tried working around it but so far unsuccessfully. So I'd default the distribution to broadcast for the union sub plans for now.

mxm commented:
@seut Pushed a fixup with a separate commit for changing the default `DistributionInfo` of `MergePhase`s to BROADCAST.

mxm commented:
Rebased to `master` and combined the fixups.
